### PR TITLE
fix: refactor Stepper.controlsBuilder to use ControlsDetails

### LINKED ISSUE #78119: Material Stepper updates button Widgets before changing Step
Refering to the material Stepper: [https://api.flutter.dev/flutter/material/Stepper-class.html](https://api.flutter.dev/flutter/material/Stepper-class.html)

The Stepper is great. However, when using `controlsBuilder` to customize the "Continue" and "Cancel" buttons, there is a terrible annoying issue that makes the whole control unusable.

The issue is as follows: When changing from one step to the next, the buttons customized are updated before the step itself changes. Therefore, for one second the new buttons flash in the screen but are placed on the previous step. So afterwards the stepper goes to the next step and they are fine. But they've flashed for a sec on the previous step, which looks ugly.

### PR DESCRIPTION
**Pull request by craiglabenz**
### Behavior

Previously, use of the `Stepper.controlsBuilder` method has the potential to create a brief flicker as the controls in the vanishing step quickly rebuild with the new `activeIndex`. An example widget using the Stepper might look like so:

```dart

/// This is updated by setState() shenanigans.
int _activeIndex = 0;

Widget build(BuildContext context) {
  return Stepper(
    controlsBuilder: (BuildContext context, {VoidCallback? onStepContinue, VoidCallback? onStepCancel}) {
      return Row(
        children: <Widget>[
          TextButton(
            onPressed: onStepContinue,
            child: Text('NEXT ${_activeIndex}'),
          ),
          TextButton(
            onPressed: onStepCancel,
            child: Text('CANCEL ${stepIndex - 1}'),
          ),
        ],
      );
    },
    steps: [...],
  );
}
```

This behavior is represented by this slowed-down gif:

---

With this fix, an alternate signature is introduced and controls built with it can make use of the new `ControlsDetails` parameter to know who _they_ are, not which step is replacing them. An example widget using the Stepper could be refactored like so:

```dart

/// This is updated by setState() shenanigans.
int _activeIndex = 0;

Widget build(BuildContext context) {
  return Stepper(
    // A new `ControlsDetails` parameter appears!
    controlsBuilder: (BuildContext context, {required ControlsDetails details }) {
      return Row(
        children: <Widget>[
          TextButton(
            onPressed: details.onStepContinue,
            // Use `stepIndex` instead of `_activeIndex`
            child: Text('NEXT ${details.stepIndex}'),
          ),
          TextButton(
            onPressed: details.onStepCancel,
            // Use `stepIndex` instead of `_activeIndex`
            child: Text('CANCEL ${details.stepIndex - 1}'),
          ),
        ],
      );
    },
    steps: [...],
  );
}
```

Rewriting such methods in this way fixes this flicker, as captured in this slowed-down gif:

### Risks

This changes the signature of a published builder function, but all external submitted tests have passed and it is not used anywhere internally.

Fixes #78119

### CODE REVIEW

File: packages/flutter/lib/src/material/stepper.dart [hunk 1]
```diff
@@ -55,6 +55,53 @@ enum StepperType {
...
+  /// If null, the 'continue' button will be disabled.
+  final VoidCallback? onStepContinue;
+
+  /// The callback called when the 'cancel' button is tapped.
+  ///
+  /// If null, the 'cancel' button will be disabled.
+  final VoidCallback? onStepCancel;
+
+  /// True if the indicated step is also the current active step.
+  bool get isActive => currentStep == stepIndex;
```
Comment (Piinks) [inline packages/flutter/lib/src/material/stepper.dart:92]: Oh I see.  😝  Reference this above for when folks are in the api docs on individual pages.

File: packages/flutter/lib/src/material/stepper.dart [hunk 2]
```diff
@@ -55,6 +55,53 @@ enum StepperType {
...
+    this.onStepCancel,
+    this.onStepContinue,
+  });
+  /// Index that is active for the surrounding Stepper widget. This may be
+  /// different from [stepIndex] if the user has just changed steps and we are
+  /// currently animating toward that step.
+  final int currentStep;
+
+  /// Index of the step for which these controls are being built. This is
+  /// not necessarily the active index, if the user has just changed steps and
```
[the remark this answers is not in the document]
Comment (craiglabenz) [reply packages/flutter/lib/src/material/stepper.dart:77]: Great suggestion.

File: packages/flutter/lib/src/material/stepper.dart [hunk 3]
```diff
@@ -55,6 +55,53 @@ enum StepperType {
   horizontal,
 }
 
+/// Container for all the information necessary to build a Stepper widget's
+/// foward and backward controls for any given step.
+///
+/// Used by [Stepper.controlsBuilder].
+@immutable
+class ControlsDetails {
+  /// Default constructor.
```
[the remark this answers is not in the document]
Comment (craiglabenz) [reply packages/flutter/lib/src/material/stepper.dart:64]: Oh nice, I didn't realize this showed up anywhere!

**Review by Piinks — APPROVED** [review]
LGTM - there is a merge conflict though

**Comment by erperejildo** [comment]
@craiglabenz I just updated to 2.8 and get this error while trying the 2nd Stepper you mentioned:
`The argument type 'Row Function(BuildContext, {required ControlsDetails details})' can't be assigned to the parameter type 'Widget Function(BuildContext, ControlsDetails)?'.`
Could you provide a full example please?

**Pull request merged**